src/server/server.ml

```diff
@@ -410,16 +410,47 @@ module FlowProgram : Server.SERVER_PROGRAM = struct
       FlowExitStatus.(exit Server_out_of_date)
     end;
 
-    (* Die if a package.json changed *)
-    let modified_packages = SSet.filter (fun f ->
+    let is_incompatible filename_str =
+      let filename = Loc.JsonFile filename_str in
+      let filename_set = FilenameSet.singleton filename in
+      let ast =
+        (* Copied these options from types_js.ml, where the other reparse callsite is.
+           It's not really ideal so I'm open to suggestions. I'm also not sure what
```

Comment:
I definitely think we want to have this gather-options-and-call-reparse block in only one place - it's full of details that'll be prone to drifting otherwise. But it should be simple enough to factor as a helper (in `types_js`, or wherever makes sense) that takes a `FilenameSet` that we can call from here, I think.

Perf-wise I'd be shocked if the time to reparse was perceptible. But if so, it should be pretty straightforward to record the fact that the ASTs left by this check in shared memory are in fact current.